test/provider/model/FinEnvironment.unittest.ts

```diff
@@ -1,31 +1,110 @@
 import {StoredApplication} from '../../../src/provider/model/Environment';
-import {createMockFin} from '../../mocks';
-
-describe('When the same app is attempted to be launched multiple times instantaneously through the Environment', () => {
-    const storedApp: StoredApplication = {
-        type: 'manifest',
-        title: 'cr test app',
-        id: 'cr-test-app',
-        manifestUrl: 'some-manifest-url'
-    };
-    let environment: FinEnvironment;
-
-    beforeEach(async () => {
-        Object.assign(global, {
-            fin: createMockFin()
-        });
-        environment = new FinEnvironment();
+import {createMockFin, createMockApplication} from '../../utils/unit/mocks';
+import {createFakeManifestStoredApplication as createFakeManifestStoredApplication, createFakeProgrammaticApplication} from '../../utils/unit/fakes';
+
+const mockFin = createMockFin();
+
+let environment: FinEnvironment;
+
+beforeEach(async () => {
+    jest.resetAllMocks();
+
+    environment = new FinEnvironment();
+});
+
+describe('When launching a manifest application', () => {
+    testLaunch(
+        mockFin.Application.createFromManifest,
+        createFakeManifestStoredApplication,
+        (application) => application.manifestUrl
+    );
+});
+
+describe('When launching a programmatic application', () => {
+    testLaunch(
+        mockFin.Application.create,
+        createFakeProgrammaticApplication,
+        (application) => application.initialOptions
+    );
+});
+
+function testLaunch<TStoredApplication extends StoredApplication, TCreateParam>(
+    mockCreate: jest.Mock<Promise<Application>, [TCreateParam]>,
+    createFakeStoredApplication: () => TStoredApplication,
+    extractCreateParameter: (param: TStoredApplication) => TCreateParam
+): void {
+    let storedApp: TStoredApplication;
+
+    beforeEach(() => {
+        storedApp = createFakeStoredApplication();
+
+        mockCreate.mockResolvedValue(createMockApplication());
     });
 
-    test('The Environment will only attempt to start the app once', async () => {
-        for (let i = 0; i < 10; ++i) {
+    test('The app is launched through the OpenFin runtime', async () => {
+        await environment.startApplication(storedApp);
+
+        expect(mockCreate).toBeCalledTimes(1);
+        expect(mockCreate).toBeCalledWith(extractCreateParameter(storedApp));
+    });
+
+    test('Subsequent attempts to launch the same application are ignored', async () => {
+        await environment.startApplication(storedApp);
+
+        for (let i = 0; i < 10; i++) {
             await environment.startApplication(storedApp);
         }
-        expect(fin.Application.createFromManifest).toBeCalledTimes(1);
-        expect(fin.Application.createFromManifest).toBeCalledWith(storedApp.manifestUrl);
+
+        expect(mockCreate).toBeCalledTimes(1);
     });
-});
+
+    test('If app launch fails, we can attempt to launch the app again', async () => {
```

Comment:
I might be missing something really obvious, but it looks like you have this exact test here twice

Reply:
✅